### PR TITLE
samples: net: echo_client: handle net if mtu

### LINKED ISSUE #22447: samples: echo_client sample breaks for UDP when larger than net if MTU
**Describe the bug**
When testing the echo_client sample on U-Blox SARA R4 modem, the UDP send breaks.  It's trying to send a packet that is larger than the modem's MTU (1024).  Other interfaces with MTU less than 1280 probably have this issue as well, but don't see it due to the random nature of the UDP send size.

Echo_client needs to take the MTU into account for UDP packet sizes.

**To Reproduce**
1. west build -b frdm_k64f -s zephyr/samples/net/sockets/echo_client -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"<echo server>\" \
  -DCONFIG_NET_IPV6=n -DCONFIG_NET_CONFIG_NEED_IPV6=n \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<modem APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<modem MCCMNO>\" \
  -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_NET_CONFIG_NEED_IPV4=y
2. Wait for a UDP packet send of > 1024.

**Expected behavior**
UDP packet send should not exceed the MTU of the network interface.  Each packet needs to be self-contained.

**Impact**
echo_client sample should work with most network interfaces (bug).

**Environment (please complete the following information):**
 - OS: Arch Linux
 - Zephyr SDK 0.11
 - Commit SHA: be0f5fe0b0be7d095419ee90d23c8831cf8de9c1 (master a few days ago)

### PR DESCRIPTION
**Pull request by mike-scott**
UDP portion of the echo_client sends 1 packet with the sample data.
(TCP can send chunks of it depending on the response of the send()
function.)  Not every network interface can send a UDP packet large
enough to handle the size of the sample data.

Let's make sure to account for the network interface MTU when deciding
the amount of sample data to send.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22447

Signed-off-by: Michael Scott <mike@foundries.io>

### CODE REVIEW

File: samples/net/sockets/echo_client/src/udp.c [hunk 1]
```diff
@@ -32,7 +32,8 @@ static int send_udp_data(struct data *data)
 
 	do {
 		data->udp.expecting = sys_rand32_get() % ipsum_len;
-	} while (data->udp.expecting == 0U);
+	} while (data->udp.expecting == 0U ||
+		 data->udp.expecting > data->udp.mtu);
```
Comment (tbursztyka) [inline samples/net/sockets/echo_client/src/udp.c]: Usually, application should never care about the MTU. A proper fix would be to loop over send() until all the data has been sent (or an error occurred).

Actually, I just checked: look at tcp.c of the same sample, it does exactly that through a local sendall() function.

Comment (mike-scott) [reply samples/net/sockets/echo_client/src/udp.c]: @tbursztyka I dont think UDP should work with the send() loop in the same way as TCP, as most protocols will need protocol header information at the beginning of each packet.  However, for echo_client it would probably work.

The question is: what's the right behavior?

Comment (mike-scott) [reply samples/net/sockets/echo_client/src/udp.c]: @tbursztyka one idea: is to save the results of `send()` back to `data->udp.expecting` (not use a loop).  This would take care of any MTU limit imposed by the network interface.

For reference, here are a few links to existing code in Zephyr for UDP-based samples/libraries:
Coap-Client: https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/coap_client/src/coap-client.c#L187
UpdateHub: https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/updatehub/updatehub.c#L306
LwM2M: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/lwm2m/lwm2m_engine.c#L1031

Comment (tbursztyka) [reply samples/net/sockets/echo_client/src/udp.c]: Theoretically you cannot use send() positive result as the MTU as it may send less. On zephyr it will send as the size of MTU because it's a simple stack but your application code should not care about the stack internals.

Btw, same logic applies to recv() (if for instance you request to read less than what has/is going to be received etc...)

If a protocol has internal optimization depending on the MTU, then that's another story (for instance when such protocol manages re-transmission of missed packet instead of re-sending the whole data all over again). But so far I don't think any of the mentioned protocol manages that (maybe I am wrong).

Anyway, here the point of echo application is to send an arbitrary amount of data, no matter the MTU size. Thus why tcp side is already doing it properly btw.

Comment (rlubos) [reply samples/net/sockets/echo_client/src/udp.c]: I don't think we should consider any kind of UDP fragmentation at the application (socket) level. UDP (DGRAM) sockets behave differently from TCP (STREAM) - you cannot `recv` datagram in chunks and I'd expect the same from `send`. While it's not clearly stated in `send` manual, it's explicit for the `recv`:
```
For message-based sockets, such as SOCK_DGRAM and SOCK_SEQPACKET, the entire message shall be read in
a  single  operation.  If a message is too long to fit in the supplied buffer, and MSG_PEEK is not set in the
flags argument, the excess bytes shall be discarded.
```
For `send` we have:
```
If the message is too long to pass through the underlying protocol, send() shall fail and no data shall
be transmitted.
```
and the following errno code specified:
```
EMSGSIZE
              The message is too large to be sent all at once, as the socket requires.
```

As for the MTU management, I'd expect this to be an IP layer responsibility to fragment the packet at IP level (shouldn't be the case for IPv6 though as the minimal MTU for this protocol is 1280 bytes). But it seems that fragmentation is not handled well as we discuss here.

One possibility to consider to relieve the application from a manual MTU management would be to handle `EMSGSIZE` returned by `send`, and simply re-roll the packet length in that case instead of failing. Up to you to decide what'd be neater.

Comment (jukkar) [reply samples/net/sockets/echo_client/src/udp.c]: As TCP is a stream protocol, we obviously need to handle partial sends. For UDP, the application normally just sends discrete packets. In order to figure out what size to use for a packet, the MTU is important information. I can also envision a system where the app would try to send large packets and then re-send the data that was not sent. So it really depends on application logic. IMHO using MTU information here is just fine.

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
> Not every network interface can send a UDP packet large
enough to handle the size of the sample data.

IP protocols have requirement for minimum MTU required. For IPv6, it's 1280 bytes. For IPv4, it's much more modest (and probably less official), 576 bytes. We just should send sample packets within these bounds, and all compliant interfaces should support them.

File: samples/net/sockets/echo_client/src/udp.c (hunk 1 above)
Comment (tbursztyka) [reply samples/net/sockets/echo_client/src/udp.c]: @rlubos I thought EMSGSIZE applied only to sendmsg since it details after, that for send() no implicit error code is ever returned, just -1. But maybe I am confused by how the man page sentences ore ordered here.

Comment (rlubos) [reply samples/net/sockets/echo_client/src/udp.c]: @tbursztyka 
Well, I agree these man pages are sometimes hard to follow. If you refer to:
```
If the message is too long to pass atomically through the underlying protocol, the error EMSGSIZE is returned, and the message is not transmitted.

No indication of failure to deliver is implicit in a send(). Locally detected errors are indicated by a return value of -1.
```
which is part of the Linux Programmer's manual (`man 2 send`), I don't think the two paragraphs are related. The second one seems to be formulated cleaner in the POSIX Programmer's manual (`man 3 send`):
```
Successful completion of a call to send() does not guarantee delivery of the message. A return  value  of  −1 indicates only locally-detected errors
```

**Comment by jukkar** [comment]
> We just should send sample packets within these bounds, and all compliant interfaces should support them.

The implementations are buggy and here we are seeing one. Not sure whether the sara modem offloading actually supports IPv6 (this bug suggests that it might not). This should not be an issue if PPP is used with this modem (the IP/TCP/UDP offloading with modem drivers are not really needed anymore because we have the PPP support with generic GSM modem).

Anyway, we can always send less than MTU and that this PR is trying to enforce. For receiving data, the link layer must provide its own fragmentation and reassembly mechanism to ensure that a 1280-byte IPv6 datagram can be delivered intact to IP layer.

**Comment by mike-scott** [comment]
To summarize, we have 2 possible resolutions to fix the `echo_client` sample for net interfaces which supports less than 1280 MTU:
- Keep this as is and use MTU directly
- Use the result of the send() as what is expected to be returned -- this leaves the max send size up to the driver.  In this case, we should also handle errno == EMSGSIZE where we should just try again with a different random number and hope it works the next time.

@pfalcon 
After re-reading parts of RFC791, I don't think we can arbitrarily limit the send size to 576 and expect that it will work on every network interface.

That value is described like this:
`All hosts must be prepared to accept datagrams of up to 576 octets (**whether they arrive whole or in fragments**).`
So, this is more of a guideline for the size of datagrams the interface should be prepared to **receive**.
The **send** is described as:
`Every internet module must be able to forward a datagram of 68 octets without further fragmentation.  This is because an internet header may be up to 60 octets, and the minimum fragment is 8 octets.`

I'm reading this as the actual sending "minimum" MTU is 68.  That's far too small of a value to use here, in my opinion.

**Comment by jukkar** [comment]
Lets go with this for now, can be changed later if needed.
